students/karisPlath/Lab1/lab1.ino

```diff
@@ -0,0 +1,103 @@
+// LED buttons
+const int BLED = 9;
+const int GLED = 10;
+const int RLED = 11;
+const int BUTTON = 2;
+
+// debounce variables
+boolean lastButton = LOW;
+boolean currentButton = LOW;
+
+// state + mode
+int ledMode = 0;
+int numState = 5;
+
+// initialize I/O
+void setup() {
+  pinMode(BLED, OUTPUT);
+  pinMode(GLED, OUTPUT);
+  pinMode(RLED, OUTPUT);
+
+ // setup interrupt button 
+  pinMode(BUTTON, INPUT_PULLUP);
+  attachInterrupt(digitalPinToInterrupt(BUTTON), interrupt, CHANGE);
+}
+
+// debounce button
+boolean debounce(boolean last){
+  boolean current = digitalRead(BUTTON);
+  if(last != current){
+    delay(5);
+    current = digitalRead(BUTTON);
+  }
+  return current;
+}
+
+// set color of LED
+void setLED(int r, int g, int b){
+  analogWrite(RLED, r);
+  analogWrite(GLED, g);
+  analogWrite(BLED, b);
+}
+
+// interrupt on button press
+void interrupt(){
+  // interrupt to debounce button
+  currentButton = debounce(lastButton);
+  if(lastButton == LOW && currentButton == HIGH){
+    // next mode
+    ledMode++; 
+    Serial.println(ledMode);
+  }
+  lastButton = currentButton;
+}
+
+void loop() {
+  // check if mode resets
+  if(ledMode == numState){
+    ledMode = 0;
+  }
+  // turn LED off
+  else if(ledMode == 0){
+    setLED(0, 0, 0);
+  }
+  // blink LED
+  else if(ledMode == 1){
+    setLED(255, 255, 255);
+    delay(500);
+    setLED(0, 0, 0);
+    delay(500);
+  }
+  // purple fade
+  else if(ledMode == 2){
+    for(int i = 0; i < 255; i++){
+      delay(10);
+      setLED(i, 0, i);
+    }
```

Comment:
Hm, even with your cheating interrupt, my guess is if you press the button the fade completes its cycle before changing. Which is probably very annoying if you want to cycle through quickly. So even better would be to write the code where it doesn't spend a lot of time in any one mode case. Have it do an update and cycle back through, all the while the interrupt can watch for a button press.